### PR TITLE
Add trailing stoploss hyperspace

### LINKED ISSUE #2363: Hyperopt trailing stoploss hyperoptimization space
## My environment

* Operating system: Windows
* Python Version: Python 3.7.4
* Branch: freqtrade develop-7d1f66cc

### The suggestion
In hyperopt I currently can't find an easy way to hyperopt values like `trailing_stop_positive` or `trailing_stop_positive_offset`. Maybe make a way that you can easily add them like the `stoploss` value.

### PR DESCRIPTION
**Pull request by hroff-1902**
Adds the new 'trailing' hyperspace for Hyperopt for optimizing trailing stoploss parameters.

Resolves #2363

* Change to the `--spaces`:
I do not think this should be in the optimization path by default. Optimization for the trailing stop parameters should better be done differently (or specified explicitly, if user wants to use it along with optimizing buy, sell, roi and stoploss). So the default value for `--spaces` was changed to `default` (new value) from `all`, and this new `trailing` hyperspace is not included into `default` set of hyperspaces.

So, use `--space all` or `--space trailing` or `--spaces defailt trailing` to switch it on.

* The `trailing_stop` setting is set to True always when the `trailing` space is specified (so it's not varied between True and False).

Usage:
```
$ freqtrade hyperopt --spaces trailing
$ freqtrade hyperopt --spaces stoploss trailing --print-json
```

<s>It requires some additional love, so PoC for now.</s>

TODO:
* [x] more tests
* [x] docs

### CODE REVIEW

**Comment by hroff-1902** [comment]
Seems to be ready, pls review

File: docs/hyperopt.md [hunk 1]
```diff
@@ -46,8 +46,9 @@ Optional - can also be loaded from a strategy:
 Rarely you may also need to override:
 
 * `roi_space` - for custom ROI optimization (if you need the ranges for the ROI parameters in the optimization hyperspace that differ from default)
-* `generate_roi_table` - for custom ROI optimization (if you need more than 4 entries in the ROI table)
+* `generate_roi_table` - for custom ROI optimization (if you need more than 4 entries in the ROI table or the ranges of the values in the ROI table that differ from default)
```
[the remark this answers is not in the document]
Comment (hroff-1902) [reply docs/hyperopt.md]: "if you need ... ranges of values ... that differ from default"

(there are many values in the ROI tables, diff. values have diff. ranges, so "ranges that differ")

I'll try to reformulate it.

File: freqtrade/optimize/hyperopt.py [hunk 2]
```diff
@@ -190,25 +192,35 @@ def log_trials_result(self) -> None:
...
+            if self.has_space('trailing'):
+                result_dict.update(self.space_params(params, 'trailing'))
+
             print(rapidjson.dumps(result_dict, default=str, number_mode=rapidjson.NM_NATIVE))
         else:
             if self.has_space('buy'):
                 print('Buy hyperspace params:')
-                pprint({p.name: params.get(p.name) for p in self.hyperopt_space('buy')},
-                       indent=4)
+                pprint(self.space_params(params, 'buy', 5), indent=4)
```
[the remark this answers is not in the document]
Comment (hroff-1902) [reply freqtrade/optimize/hyperopt.py]: Rafactored, pls review

File: freqtrade/optimize/hyperopt.py [hunk 3]
```diff
@@ -170,57 +170,43 @@ def log_trials_result(self) -> None:  # noqa: C901
...
-                pprint(self.space_params(params, 'stoploss', 5), indent=4)
-
-            if self.has_space('trailing'):
-                print('Trailing stop:')
-                pprint(self.space_params(params, 'trailing', 5), indent=4)
+    def _params_pretty_print(self, params, space: str, header: str):
+        if self.has_space(space):
+            space_params = self.space_params(params, space, 5)
+            print(header)
+            pprint(space_params, indent=4)
```
Comment (xmatthias) [inline freqtrade/optimize/hyperopt.py]: Now i have to confess, this looks pretty neat :+1:

Comment (hroff-1902) [reply freqtrade/optimize/hyperopt.py]: it was cleaner for a reader before, imho

**Pull request merged**